crates/net/eth-wire/src/p2pstream.rs

```diff
@@ -613,19 +613,23 @@ where
     /// Returns `Poll::Ready(Ok(()))` when no buffered items remain.
     fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
         let mut this = self.project();
-        loop {
-            match ready!(this.inner.as_mut().poll_flush(cx)) {
-                Err(err) => return Poll::Ready(Err(err.into())),
-                Ok(()) => {
-                    let Some(message) = this.outgoing_messages.pop_front() else {
-                        return Poll::Ready(Ok(()))
-                    };
-                    if let Err(err) = this.inner.as_mut().start_send(message) {
-                        return Poll::Ready(Err(err.into()))
-                    }
-                }
+        let res = loop {
+            // P2PStream wraps `reth_ecies::stream::ECIESStream` in node. `ECIESStream`
+            // internally wraps the network connection in `tokio_util::codec::Framed`.
+            //
+            // Call to `Framed::poll_ready` will flush internally, if limit for buffered messages
+            // reached. `Framed::poll_ready`, only returns `Poll::Pending` or `Poll::Ready(Err(e))`
```

Comment:
this is true, but this means we don't necessarily flush here?

before calling flush the caller already ensured that this sink is ready to flush, so doesn't seem right to me to do this again here

Reply:
it's advised by the docs to call `poll_ready` before each call to `start_send`

> Begin the process of sending a value to the sink. Each call to this function must be preceded by a successful call to poll_ready which returned Poll::Ready(Ok(())).

https://docs.rs/futures/latest/futures/sink/trait.Sink.html#tymethod.start_send

I can see how it's necessary, because `FramedImpl` controls frame size. if its buffer is too full, it calls `poll_flush` as part of its`poll_ready` impl

Reply:
this is still the wrong order here, 

```
loop {
 ready
 send
}
```

does not seem appropriate here

Reply:
pls elaborate

Reply:
calling ready+send here again before doing flush is not required because the caller already did that before calling flush, but I agree that the current impl violates the sink semantics

Reply:
to respect the fram size, of underlying `Frame` impl that we use in `P2PStream` in `reth-network`, we must call `poll_ready` before `start_send`, as mentioned in the comment https://docs.rs/tokio-util/latest/src/tokio_util/codec/framed_impl.rs.html#262-268

Reply:
but this is already done before calling flush so doing it again in flush seems pointless to me

Reply:
sure, but for `Framed` to decide the frame size, we need to call it again before every call to its`start_send` method, so that the `Framed` interface can decide, if so many items have been queued with the previous calls to `start_send`, that it is time for `Framed` to flush again

Reply:
then this isn't the right fix
because we only need to call flush then

Reply:
`poll_ready` on `Framed` calls flush